File: c/tests/test_ecdh.c
/*
Copyright 2015 CertiVox UK Ltd

This file is part of The CertiVox MIRACL IOT Crypto SDK (MiotCL)

MiotCL is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

MiotCL is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with MiotCL.  If not, see <http://www.gnu.org/licenses/>.

You can be released from the requirements of the license by purchasing 
a commercial license.
*/

/* test driver and function exerciser for ECDH/ECIES/ECDSA API Functions */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include "ecdh.h"

int main()
{   
  int i,j=0,res;
  int result;
  unsigned long ran;
  char *pp="M0ng00se";
  /* These octets are automatically protected against buffer overflow attacks */
  /* Note salt must be big enough to include an appended word */
  /* Note ECIES ciphertext C must be big enough to include at least 1 appended block */
  /* Recall EFS is field size in bytes. So EFS=32 for 256-bit curve */
  char s0[EGS],s1[EGS],w0[2*EFS+1],w1[2*EFS+1],z0[EFS],z1[EFS],raw[100],key[EAS],salt[32],pw[20],p1[30],p2[30],v[2*EFS+1],m[32],c[64],t[32],cs[EGS],ds[EGS];
  octet S0={0,sizeof(s0),s0};
  octet S1={0,sizeof(s1),s1};
  octet W0={0,sizeof(w0),w0};
  octet W1={0,sizeof(w1),w1};
  octet Z0={0,sizeof(z0),z0};
  octet Z1={0,sizeof(z1),z1};
  octet RAW={0,sizeof(raw),raw};
  octet KEY={0,sizeof(key),key};
  octet SALT={0,sizeof(salt),salt};
  octet PW={0,sizeof(pw),pw};
  octet P1={0,sizeof(p1),p1};
  octet P2={0,sizeof(p2),p2};
  octet V={0,sizeof(v),v};
  octet M={0,sizeof(m),m};
  octet C={0,sizeof(c),c};
  octet T={0,sizeof(t),t};
  octet CS={0,sizeof(cs),cs};
  octet DS={0,sizeof(ds),ds};

  /* Crypto Strong RNG */
  csprng RNG;                  				  
  time((time_t *)&ran);
  /* fake random seed source */
  RAW.len=100;		
  RAW.val[0]=ran;
  RAW.val[1]=ran>>8;
  RAW.val[2]=ran>>16;
  RAW.val[3]=ran>>24;
  for (i=0;i<100;i++) RAW.val[i]=i;
  /* initialise strong RNG */
  CREATE_CSPRNG(&RNG,&RAW);   

  SALT.len=8;
  for (i=0;i<8;i++) SALT.val[i]=i+1;  // set Salt

  printf("Alice's Passphrase= %s\n",pp);

  OCT_empty(&PW);
  OCT_jstring(&PW,pp);   // set Password from string

  /* private key S0 of size EGS bytes derived from Password and Salt */
  PBKDF2(&PW,&SALT,1000,EGS,&S0);
  printf("Alices private key= 0x"); OCT_output(&S0);

  /* Generate Key pair S/W */
  ECP_KEY_PAIR_GENERATE(NULL,&S0,&W0); 

  res=ECP_PUBLIC_KEY_VALIDATE(1,&W0);
  if (res!=0)
  {
    printf("ECP Public Key is invalid!\n");
    return 1;
  }

  printf("Alice's public key= 0x");  OCT_output(&W0);

  /* Random private key for other party */
  ECP_KEY_PAIR_GENERATE(&RNG,&S1,&W1);
  res=ECP_PUBLIC_KEY_VALIDATE(1,&W1);
  if (res!=0)
  {
    printf("ECP Public Key is invalid!\n");
    return 1;
  }
  printf("Servers private key= 0x");  OCT_output(&S1);
  printf("Servers public key= 0x");   OCT_output(&W1);

  /* Calculate common key using DH - IEEE 1363 method */
  ECPSVDP_DH(&S0,&W1,&Z0);
  ECPSVDP_DH(&S1,&W0,&Z1);
  
  if (!OCT_comp(&Z0,&Z1))
  {
    printf("*** ECPSVDP-DH Failed\n");
    return 0;
  }

  KDF2(&Z0,NULL,EAS,&KEY);

  printf("Alice's DH Key=  0x"); OCT_output(&KEY);
  printf("Servers DH Key=  0x"); OCT_output(&KEY);

  printf("Testing ECIES\n");

  P1.len=3; P1.val[0]=0x0; P1.val[1]=0x1; P1.val[2]=0x2; 
  P2.len=4; P2.val[0]=0x0; P2.val[1]=0x1; P2.val[2]=0x2; P2.val[3]=0x3; 

  M.len=17;
  for (i=0;i<=16;i++) M.val[i]=i; 

  ECP_ECIES_ENCRYPT(&P1,&P2,&RNG,&W1,&M,12,&V,&C,&T);

  printf("Ciphertext= \n"); 
  printf("V= 0x"); OCT_output(&V);
  printf("C= 0x"); OCT_output(&C);
  printf("T= 0x"); OCT_output(&T);

  if (!ECP_ECIES_DECRYPT(&P1,&P2,&V,&C,&T,&S1,&M))
  {
    printf("*** ECIES Decryption Failed\n");
    return 1;
  }
  else printf("Decryption succeeded\n");

  printf("Message is 0x"); OCT_output(&M);

  printf("Testing ECDSA\n");

  if (ECPSP_DSA(&RNG,&S0,&M,&CS,&DS)!=0)
  {
    printf("***ECDSA Signature Failed\n");
    return 1;
  }

  printf("Signature C = 0x"); OCT_output(&CS);
  printf("Signature D = 0x"); OCT_output(&DS);

  if (ECPVP_DSA(&W0,&M,&CS,&DS)!=0)
  {
    printf("***ECDSA Verification Failed\n");
    return 1;
  }
  else printf("ECDSA Signature/Verification succeeded %d\n",j);

  KILL_CSPRNG(&RNG);

  printf("SUCCESS\n");
  return 0;
}

File: swift/TestECDH.swift
//
//  TestECDH.swift
//  miot2
//
//  Created by Michael Scott on 02/07/2015.
//  Copyright (c) 2015 Michael Scott. All rights reserved.
//

import Foundation
import clint

public func TestECDH()
{
    var pp=String("M0ng00se");
    
    let EGS=ECDH.EGS
    let EFS=ECDH.EFS
    let EAS=AES.KS
    
    var S1=[UInt8](count:EGS,repeatedValue:0)
    var W0=[UInt8](count:2*EFS+1,repeatedValue:0)
    var W1=[UInt8](count:2*EFS+1,repeatedValue:0)
    var Z0=[UInt8](count:EFS,repeatedValue:0)
    var Z1=[UInt8](count:EFS,repeatedValue:0)
    var RAW=[UInt8](count:100,repeatedValue:0)
    var SALT=[UInt8](count:8,repeatedValue:0)
    var P1=[UInt8](count:3,repeatedValue:0)
    var P2=[UInt8](count:4,repeatedValue:0)
    var V=[UInt8](count:2*EFS+1,repeatedValue:0)
    var M=[UInt8](count:17,repeatedValue:0)
    var T=[UInt8](count:12,repeatedValue:0)
    var CS=[UInt8](count:EGS,repeatedValue:0)
    var DS=[UInt8](count:EGS,repeatedValue:0)
    
    var rng=RAND()
    
    rng.clean();
    for var i=0;i<100;i++ {RAW[i]=UInt8(i&0xff)}
    
    rng.seed(100,RAW)
    
    
    for var i=0;i<8;i++ {SALT[i]=UInt8(i+1)}  // set Salt
    
    println("Alice's Passphrase= "+pp)
    var PW=[UInt8](pp.utf8)
    
    /* private key S0 of size EGS bytes derived from Password and Salt */
    
    var S0=ECDH.PBKDF2(PW,SALT,1000,EGS)
    print("Alice's private key= 0x"); ECDH.printBinary(S0)
    
    /* Generate Key pair S/W */
    ECDH.KEY_PAIR_GENERATE(nil,&S0,&W0);
    
    print("Alice's public key= 0x"); ECDH.printBinary(W0)
    
    var res=ECDH.PUBLIC_KEY_VALIDATE(true,W0);

    if res != 0
    {
        println("ECP Public Key is invalid!");
        return;
    }
    
    /* Random private key for other party */
    ECDH.KEY_PAIR_GENERATE(rng,&S1,&W1)
    
    print("Servers private key= 0x"); ECDH.printBinary(S1)
    
    print("Servers public key= 0x"); ECDH.printBinary(W1);
    
    res=ECDH.PUBLIC_KEY_VALIDATE(true,W1)
    if res != 0
    {
        println("ECP Public Key is invalid!")
        return
    }
    
    /* Calculate common key using DH - IEEE 1363 method */
    
    ECDH.ECPSVDP_DH(S0,W1,&Z0)
    ECDH.ECPSVDP_DH(S1,W0,&Z1)
    
    var same=true
    for var i=0;i<EFS;i++
    {
        if Z0[i] != Z1[i] {same=false}
    }
    
    if (!same)
    {
        println("*** ECPSVDP-DH Failed")
        return
    }
    
    var KEY=ECDH.KDF1(Z0,EAS)
    
    print("Alice's DH Key=  0x"); ECDH.printBinary(KEY)
    print("Servers DH Key=  0x"); ECDH.printBinary(KEY)

    println("Testing ECIES")
    
    P1[0]=0x0; P1[1]=0x1; P1[2]=0x2
    P2[0]=0x0; P2[1]=0x1; P2[2]=0x2; P2[3]=0x3
    
    for var i=0;i<=16;i++ {M[i]=UInt8(i&0xff)}
    
    var C=ECDH.ECIES_ENCRYPT(P1,P2,rng,W1,M,&V,&T)

    println("Ciphertext= ")
    print("V= 0x"); ECDH.printBinary(V)
    print("C= 0x"); ECDH.printBinary(C)
    print("T= 0x"); ECDH.printBinary(T)

    M=ECDH.ECIES_DECRYPT(P1,P2,V,C,T,S1)
    if M.count==0
    {
        println("*** ECIES Decryption Failed\n")
        return
    }
    else {println("Decryption succeeded")}
    
    print("Message is 0x"); ECDH.printBinary(M)
    
    println("Testing ECDSA")

    if ECDH.ECPSP_DSA(rng,S0,M,&CS,&DS) != 0
    {
        println("***ECDSA Signature Failed")
        return
    }
    println("Signature= ")
    print("C= 0x"); ECDH.printBinary(CS)
    print("D= 0x"); ECDH.printBinary(DS)
    
    if ECDH.ECPVP_DSA(W0,M,CS,DS) != 0
    {
        println("***ECDSA Verification Failed")
        return
    }
    else {println("ECDSA Signature/Verification succeeded ")}

}

TestECDH()
